edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/M2Mqtt/M2MqttClient.cs

```diff
@@ -0,0 +1,273 @@
+            var qoses = new List<byte>();
+            foreach (var entry in subscriptions)
+            {
+                topics.Add(entry.Key);
+                qoses.Add(Convert.ToByte((int)entry.Value));
+            }
+
+            async Task<Dictionary<string, Qos>> subscribeTaskSupplier()
+            {
+                lock (_lock)
+                {
+                    if (!cancellationToken.IsCancellationRequested)
+                    {
+                        var subscribeId = _client.Subscribe(topics.ToArray(), qoses.ToArray());
+                        _pendingSubscribes.TryAdd(subscribeId, source);
+                    }
+                }
+
+                var subscribeResult = await source.Task;
+                if (subscribeResult.Length != topics.Count)
+                {
+                    throw new MqttException(message: $"MQTT server did not grant QoS for every requested subscription, expected={subscriptions.Count}, but was {subscribeResult.Length}.");
+                }
+
+                var gainedQoses = new Dictionary<string, Qos>();
+                for (var i=0; i<subscribeResult.Length; i++)
+                {
+                    var gained = Convert.ToInt32(subscribeResult[i]);
+                    if (gained < 0 || gained > 2)
+                    {
+                        throw new MqttException(message: $"Subscribe [topiic={topics[i]}, qos={qoses[i]} failed: [resultCode={gained}].");
+                    }
+                    gainedQoses[topics[i]] = (Qos)gained;
+                }
+
+                return gainedQoses;
+            }
+
+            Task restoreTaskSupplier()
+            {
+                source.TrySetCanceled();
+                return Task.FromResult(true);
+            }
+
+            return RunTaskWithCancellationTokenAsync(subscribeTaskSupplier, restoreTaskSupplier, cancellationToken);
+        }
+
+        public Task UnsubscribeAsync(IEnumerable<string> topics, CancellationToken cancellationToken)
+        {
+            Preconditions.CheckNotNull(topics, nameof(topics));
+            Preconditions.CheckArgument(topics.Count() > 0, "No topics.");
+
+            var source = new TaskCompletionSource<bool>();
+            Task<bool> unsubscribeTaskSupplier()
+            {
+                lock (_lock)
+                {
+                    if (!cancellationToken.IsCancellationRequested)
+                    {
+                        var subscribeId = _client.Unsubscribe(topics.ToArray());
+                        _pendingVoidTasks.TryAdd(subscribeId, source);
+                    }
+                }
+
+                return source.Task;
+            }
+
+            Task restoreTaskSupplier()
+            {
+                source.TrySetCanceled();
+                return Task.FromResult(true);
+            }
+
+            return RunTaskWithCancellationTokenAsync(unsubscribeTaskSupplier, restoreTaskSupplier, cancellationToken);
+        }
```

Comment:
Seems that we lose the information if the QoS1 message was actually sent out (the OnMessagePublished() below sets a Task<bool> according to the eventArgs.IsPublished flag, but then it is not used here, only bare Task. As a result, the the MqttBorkerConnector::SendAsync() will not notice if IsPusblished was false, and it returns true.

Reply:
Good catch. IT should return `bool`.